### ISSUE TITLE
Use F-Droid update.zip

### ISSUE THREAD

**Issue by patcon** [patcon/mission-impossible-android#29]
This doesn't yet exist, but is apparently a work in progress. See here:
https://github.com/f-droid/fdroid-client/blob/master/F-Droid/tools/build-zip

Not sure exactly how best to use it, but seems better to pass off the f-droid system app installation to them. Perhaps we could help them move that along at some point.

cc: @eighthave

**Comment by eighthave** [comment]
the "system app" mode does work now, and you can build and flash an update.zip
to get a working FDroid as system app.  The code was even audited.  The last
remaining thing to do before FDroid's system app mode is ready for general use
is to implement a GUI that displays the permissions diff when upgrading an
installed APK, and to show the permissions that a new app is requesting.

**Comment by SchnWalter** [comment]
Hello,
Here are a set of instructions on how to use that script: https://f-droid.org/wiki/page/Installing_F-Droid_as_a_system-app#Install_using_the_flashable_zip

P.S. I've started working on updating the mission-impossible-andoid (MIA) bash scripts (see pull requtes), but I eventually realized that python is more suited for this kind of things, so i've started on updating, extending and cleaning up the fdroidserver scripts - I didn't want to start from scratch... but it will take some time utill I finish updating them.

P.S.2. I was also thinking that we could also have MIA as a set of configurable templates with different settings (since not everyone would want the most strict rules) and have MIA work like on the lines of veewee project that is used for generating vagrant templates.

**Comment by eighthave** [comment]
Just to clarify things, in case they were not apparent: the idea is to install FDroid as a system app, then it has all the same abilities as Google Play. Right now, FDroid uses an Android provided non-system-app install mode, that provides the GUI for displaying the perms stuff.  So when you install an app using FDroid in system app mode, it does not show you any screen to confirm whether you agree to the permissions that the app is claiming.  FDroid needs to have its own GUI for that stuff when running in system app mode.

**Comment by SchnWalter** [comment]
Yes, I saw that, but i was think of updating the fdroid scripts so that we can use them with MIA in order to generate the proper update.zip, not only for system apps.

And I'm very fond of the idea of different templates that you could use to generate multiple deffinitions that can later each be converted into update.zip. And people can choose what they want out of the available templates... take a look at this branch: https://github.com/SchnWalter/mission-impossible-android/tree/templates

Sorry, I know it's a bit out of topic.

**Comment by patcon** [comment]
That's awesome @SchnWalter! Feel free to start another issue or pull request if you're looking to knock ideas around about that

**Comment by SchnWalter** [comment]
I've just created a new issue: #30

**Comment by patcon** [comment]
Ref: https://f-droid.org/wiki/page/Super-F-Droid

**Comment by patcon** [comment]
> So when you install an app using FDroid in system app mode, it does not show you any screen to confirm whether you agree to the permissions that the app is claiming.

@eighthave Just to reiterate, since I still don't have clarity that we're understanding each other: MIA installs fdroid in a system app location, but we do NOT enable the expert fdroid setting for going into "system app mode". Even now, I use it to install apps, packageInstaller permissions dialogs work fine, and I have NOT enabled "Unknown sources" apks.

**Comment by SchnWalter** [comment]
I've just tested the script on a fresh devices a couple of minutes a go; I followed the steps explained in this PR #74 in order to install the OS and F-Droid.

And here is my experience:
* Opened the F-Droid application.
* Clicked the refresh button and waited to the list of apps to show up.
* Looked into the F-Droid preferences and the "Expert" options were unchecked.
* Looked for an application to install (one which requires some permissions).
* Clicked install and I saw an `Install blocked` message.
* Checked the expert options and checked the "Install using root access" - and I allowed root access for one time.
* Tried to install the application again and 
* Tried to install the application again and I was __NOT__ asked to review permissions but I was asked to give root access - allowed one time.
* Uninstalled the application.
* Tried to check the "Install using system-permissions" option.... and it didn't work, because the current version of the script has an issue and longer puts the F-Droid apk into the `/system/priv-apps` folder.
* Unchecked all "expert" options from the F-Droid preferences and then went to "System settings > Security" and checked "Unknown sources".
* Tried to install the application again and this time I saw a list of permissions before clicking the "Install" button on that screen and the installation went well.
* Uninstalled the application again.
* Closed F-Droid.

Enabled Developer tools, USB debugging with root for ADB; Launched ADB and `adb root`, `adb remount`, moved fdroid.apk to `/system/priv-apps`, `adb reboot`

And... I've tried for some time now to get Orbot to connect to the Tor network again... and it just isn't working ... :disappointed: I'll get back with the rest once I get the internet connection going.

**Comment by patcon** [comment]
That's helpful @SchnWalter. Thanks. That's unexpected based on what I see on my device, but I'll walk through and see what I get.

**Comment by eighthave** [comment]
I've gotten what @SchnWalter reported in my experience.  Something odd is
going on with your device, I'll be interested to see if we can find what it is.

**Comment by SchnWalter** [comment]
With F-Droid installed in `/system/priv-app` the permissions overview screen (or how is it called) still appears when you try to install an application. But if you check the "Expert > Install using system-permissions" or "Expert > Install using root access" the permissions overview screen goes away.

**Comment by patcon** [comment]
Thanks. And that final observation is without "unknown sources" enabled?

**Comment by eighthave** [comment]
Right now, fdroid won't try to use the system-app privileges unless the
preference is checked.  I think you found what the difference that @patcon was
seeing.

In the future, we'll make FDroid automatically use the system-app privs when
it is installed as a system-app.
https://gitlab.com/fdroid/fdroidclient/issues/69

**Comment by SchnWalter** [comment]
Yes, by default when you have a fresh install the "Unknown sources" is not checked.

And the "Unknown sources" was not checked when I tried the "Expert > *" options.

**Comment by patcon** [comment]
@SchnWalter :+1:

@eighthave But... ok. One last time, if we...

- install fdroid in `/system/priv-app/`
- leave "Unknown sources" unchecked

...then the "permissions overview screen" appears on app install/update (ie. don't need custom package installer for perms)

So.... don't we have everything we need in fdroid now, forget about the custom package installer? I think that that was what I was trying to suggest way back when -- that F-Droid is already perfect for safely installing apps conveniently without "Unknown sources" allowed. We can still custom roll our own package installer (for unattended updates, etc.), but it's not required for the main use-cases (appeasing whispersystems, rolling F-Droid into MIA, etc)

Am I summarizing this correctly?

**Comment by eighthave** [comment]
My memory here is foggy, so I'm guessing.  I think that if `PackageInstaller` is respecting the system-app privs of FDroid and allowing installs via FDroid without "Unknown Sources" being enabled, then I think that is a specific behavior of the ROM you are using.  That means it is not guaranteed to work on all ROMs or devices.

**Comment by eighthave** [comment]
Really, you should be asking @dschuermann since he wrote most of the system app support in FDroid.

**Comment by dschuermann** [comment]
As just explained on [F-Droid's issue 69](https://gitlab.com/fdroid/fdroidclient/issues/69), the root and system install expert options are currently not prime time ready because (as you already noticed) no permission screen is shown. As a matter of fact, the only advantage in using root commands or internal install APIs (via system permissions) is to get rid of the permission screen.
BUT: We don't want to get rid of it in all cases. So I prepared a table:

Permission screen is shown:

| case | normal install API | what we want |
|---------------------- | ---------  | --------------- |
| install new app | Y | Y |
| update app and new permissions are introduced | Y | Y (only diffs) |
| update app and no new permissions are introduced | Y | N |

So what needs to be done in F-Droid: Implement our own permission screen that is shown as defined in the table. This would also ease the process of updating many apps at once were no new permissions are introduced.

BTW: This is exactly what Google Play does. They are using the internal APIs and have their own permission screen. They even go that far and hide some permission (Internet permission) and group permissions into categories (https://support.google.com/googleplay/answer/6014972?hl=en).

Conclusively, the expert options shouldn't be used currently.

CC @mvdan
